### PR TITLE
[SPARK-26859][SQL][BACKPORT-2.3] Fix field writer index bug in non-vectorized ORC deserializer

### PR DESCRIPTION
**Pull request by manuzhang**
## What changes were proposed in this pull request?

Back port of https://github.com/apache/spark/pull/23766 to branch-2.3

## How was this patch tested?

Added UT

### CODE REVIEW

**Comment by manuzhang** [comment]
cc @dongjoon-hyun @viirya @cloud-fan

**Comment by dongjoon-hyun** [comment]
ok to test

**Comment by dongjoon-hyun** [comment]
@manuzhang Do you need this at 2.3.4 ?
cc @kiszk for Apache Spark 2.3.4.

**Comment by viirya** [comment]
The PR title is truncated. Could you fix it?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala [hunk 1]
```diff
@@ -178,6 +178,39 @@ abstract class OrcSuite extends OrcTest with BeforeAndAfterAll {
       checkAnswer(spark.read.orc(path.getCanonicalPath), Row(ts))
     }
   }
+
+  test("SPARK-26859 Append column into middle") {
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala]: I feel that it is not good to just copy the test name from original PR. That test name has its context as it is in AddColumnIntoTheMiddleTest.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala [hunk 2]
```diff
@@ -47,18 +47,20 @@ class OrcDeserializer(
...
-    var i = 0
-    while (i < validColIds.length) {
-      val value = orcStruct.getFieldValue(validColIds(i))
-      if (value == null) {
-        resultRow.setNullAt(i)
-      } else {
-        fieldWriters(i)(value)
+    var fieldWriterIndex = 0
+    var targetColumnIndex = 0
+    while (targetColumnIndex < requestedColIds.length) {
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala]: Why don't change `fieldWriters` following original PR?

**Comment by manuzhang** [comment]
@dongjoon-hyun yes, that will be great
@viirya I picked up the wrong commits. The source changes are the same as the original now and I also updated the test title.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to branch-2.3.
Thank you, @manuzhang and @viirya .

**Pull request closed without merging**